### DISCUSSION TITLE
Triton 1.0.0 required for sparse attention but does not exist

### DISCUSSION THREAD

**Discussion by maxmaier59** [deepspeedai/DeepSpeed#2408] [Q&A]
**For sparse attention Triton 1.0.0 is required:**
Only this version works with DeepSpeed

DeepSpeed C++/CUDA extension op report
--------------------------------------------------
NOTE: Ops not installed will be just-in-time (JIT) compiled at
      runtime if needed. Op compatibility means that your system
      meet the required dependencies to JIT install the op.
--------------------------------------------------
JIT compiled ops requires ninja
ninja .................. [OKAY]
--------------------------------------------------
op name ................ installed .. compatible
--------------------------------------------------
cpu_adam ............... [NO] ....... [OKAY]
cpu_adagrad ............ [NO] ....... [OKAY]
fused_adam ............. [NO] ....... [OKAY]
fused_lamb ............. [NO] ....... [OKAY]
 [WARNING]  please install triton==1.0.0 if you want to use sparse attention
sparse_attn ............ [NO] ....... [NO]
transformer ............ [NO] ....... [OKAY]
stochastic_transformer . [NO] ....... [OKAY]
async_io ............... [NO] ....... [OKAY]
utils .................. [NO] ....... [OKAY]
quantizer .............. [NO] ....... [OKAY]
transformer_inference .. [NO] ....... [OKAY]

**However, Trition 1.0.0 does not exist:**
$ pip install triton==1.0.0
ERROR: Could not find a version that satisfies the requirement triton==1.0.0 (from versions: 0.1, 0.1.1, 0.1.2, 0.1.3, 0.2.0, 0.2.1, 0.2.2, 0.2.3, 0.3.0)
ERROR: No matching distribution found for triton==1.0.0

What is the solution???

**Comment by ghost** [comment]
Requires clang and llvm 11, specifically.

```bash
#!/usr/bin/env bash

# Downloads and unzips Triton source code
wget -O triton-1.0.zip https://github.com/openai/triton/archive/refs/tags/v1.0.zip
unzip triton-1.0.zip
rm triton-1.0.zip
mv triton-1.0 triton

# Patches triton so it compiles
cd triton
{ head -n 9 include/triton/tools/graph.h; echo "using std::size_t;"; tail -n +10 include/triton/tools/graph.h; } > include/triton/tools/graph.new.h
mv include/triton/tools/graph.new.h include/triton/tools/graph.h

# Installs triton
cd python
pip3 install -e .
```

**Reply by ulysses500** [reply]
Also for windows:
Download zip: https://github.com/openai/triton/archive/refs/tags/v1.0.zip
unzip it

then, in a windows command:
cd triton/python;
pip install cmake; # build time dependency
pip install -e .
